### PR TITLE
Ensure Angle.__str__ does not assume array2print passes subclasses.

### LINKED ISSUE #7214: TST: New failures in "numpy-dev" job?
I don't see how these are related to the `einsum` thingy that is known. Example: https://travis-ci.org/astropy/astropy/jobs/342443502

```
_________________________________ test__str__ __________________________________
        # non-scalar array angles
        arrangle = Angle(['10.2345d', '-20d'])
>       strarrangle = arrangle.__str__()
astropy/coordinates/tests/test_angles.py:861: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
x = 10.2345
>   return np.array2string(self, formatter={'all': lambda x: x.to_string()})
E   AttributeError: 'numpy.float64' object has no attribute 'to_string'
astropy/coordinates/angles.py:433: AttributeError
```

```
_______________________________ test_repr_latex ________________________________
        # and array angles
        arrangle = Angle([1, 2.1], u.deg)
>       rlarrangle = arrangle._repr_latex_()
astropy/coordinates/tests/test_angles.py:880: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
x = 1.0
>   return np.array2string(self, formatter={'all': lambda x: x.to_string(format='latex')})
E   AttributeError: 'numpy.float64' object has no attribute 'to_string'
astropy/coordinates/angles.py:441: AttributeError
```

```
______________________ TestTableFunctions.test_mask_array ______________________
        expect = r"\[\('NGC1002',\s+12.3[0-9]*\) \(\'NGC1003\',\s+15.[0-9]+\)\]"
>       assert re.match(expect, str(hdu.data))
E       assert None
E        +  where None = <function match at 0x7f9c2068fe18>("\\[\\('NGC1002',\\s+12.3[0-9]*\\) \\(\\'NGC1003\\',\\s+15.[0-9]+\\)\\]", "[(b'NGC1002', 12.3) (b'NGC1003', 15.2)]")
E        +    where <function match at 0x7f9c2068fe18> = re.match
E        +    and   "[(b'NGC1002', 12.3) (b'NGC1003', 15.2)]" = str(FITS_rec([(b'NGC1002', 12.3), (b'NGC1003', 15.2)],\n         dtype=(numpy.record, [('target', 'S20'), ('V_mag', '>f4')])))
E        +      where FITS_rec([(b'NGC1002', 12.3), (b'NGC1003', 15.2)],\n         dtype=(numpy.record, [('target', 'S20'), ('V_mag', '>f4')])) = <astropy.io.fits.hdu.table.BinTableHDU object at 0x7f9bb7bf6438>.data
astropy/io/fits/tests/test_table.py:860: AssertionError
```

```
_____________________ [doctest] docs/coordinates/index.rst _____________________
248 :func:`astropy.coordinates.EarthLocation.get_site_names`.
249 
250 For arbitrary Earth addresses (e.g., not observatory sites), use the
251 `~astropy.coordinates.EarthLocation.of_address` classmethod. Any address passed
252 to this function uses Google maps to retrieve the latitude and longitude and can
253 also (optionally) query Google maps to get the height of the location. As with
254 Google maps, this works with fully specified addresses, location names, city
255 names, and etc.::
256 
257     >>> EarthLocation.of_address('1002 Holy Grail Court, St. Louis, MO')  # doctest: +REMOTE_DATA +FLOAT_CMP
Expected:
    <EarthLocation (-26726.98216371, -4997009.8604809, 3950271.16507911) m>
Got:
    <EarthLocation (-26724.93406367, -4997006.87829221, 3950274.92599205) m>
/tmp/astropy-test-eem6lc45/lib/python3.6/site-packages/docs/coordinates/index.rst:257: DocTestFailure
```

```
_________________ [doctest] docs/io/fits/usage/unfamiliar.rst __________________
020 same, i.e. the data is in the ``.data`` attribute and the ``field()`` method
021 is used to refer to the columns and returns a numpy array. When reading the
022 table, Astropy will automatically detect what kind of table it is.
023 
024 ::
025 
026     >>> from astropy.io import fits
027     >>> filename = fits.util.get_testdata_filepath('ascii.fits')
028     >>> hdul = fits.open(filename)
029     >>> hdul[1].data[:1]  # doctest: +FLOAT_CMP
Expected:
    FITS_rec([(10.123, 37)],
             dtype=(numpy.record, {'names':['a','b'], 'formats':['S10','S5'], 'offsets':[0,11], 'itemsize':16}))
Got:
    FITS_rec([(b'.10123E+02', b'   37')],
             dtype=(numpy.record, {'names':['a','b'], 'formats':['S10','S5'], 'offsets':[0,11], 'itemsize':16}))
/tmp/astropy-test-eem6lc45/lib/python3.6/site-packages/docs/io/fits/usage/unfamiliar.rst:29: DocTestFailure
```

### PR DESCRIPTION
**Pull request by mhvk**
In numpy 1.14.1, fixes to array2print include a cast of all input to a regular ndarray. Hence, in formatters one can no longer assume that what is passed on is a subclass. To counteract this, we now explicitly cast back to Angle in __str__ and _repr_latex.

fixes #7214 

(Note: the bug affects 2.0 too, hence the 2.0.5 milestone)

### CODE REVIEW

File: astropy/coordinates/angles.py [hunk 1]
```diff
@@ -426,19 +427,26 @@ def is_within_bounds(self, lower=None, upper=None):
             ok &= np.all(self < Angle(upper))
         return bool(ok)
 
-    def __str__(self):
+    def _str_helper(self, format=None):
         if self.isscalar:
-            return str(self.to_string())
```
Comment (mhvk) [inline astropy/coordinates/angles.py]: I removed the `str(...)` in my version, which is definitely OK for everything we support in astropy 3.0, but may need checking for backporting.

**Comment by astrofrog** [comment]
The Travis failure is due to a bug in Cython that we've worked around in https://github.com/astropy/astropy/pull/7232

**Comment by astrofrog** [comment]
@mhvk - can you rebase now that I've merged https://github.com/astropy/astropy/pull/7232?

**Comment by astrofrog** [comment]
@mhvk - thanks! This helps, but there are still to test failures - do you have any idea how to fix them?

**Comment by mhvk** [comment]
Hmm, yes, those may well be related. Will need to have a closer look.

**Comment by drdavella** [comment]
@mhvk I took the liberty of pushing to this branch to xfail/skip the failing cases in FITS. @astrofrog, @bsipocz, @pllim if/when CI passes, can someone review so that we can merge and then rebase other failing PRs?

File: docs/io/fits/usage/unfamiliar.rst [hunk 2]
```diff
@@ -26,6 +26,8 @@ table, Astropy will automatically detect what kind of table it is.
     >>> from astropy.io import fits
     >>> filename = fits.util.get_testdata_filepath('ascii.fits')
     >>> hdul = fits.open(filename)
+
+.. doctest-requires:: numpy<=1.14.1
```
Comment (drdavella) [inline docs/io/fits/usage/unfamiliar.rst]: Apparently this doesn't actually work, because I got the condition wrong but the test is skipped anyway...

Comment (drdavella) [reply docs/io/fits/usage/unfamiliar.rst]: I should probably just change it to `.. doctest-skip` for now. It's looking for a module called `numpy<=1.14.1` which of course is going to cause it to be skipped.

Comment (pllim) [reply docs/io/fits/usage/unfamiliar.rst]: So is that a bug or feature? If a bug, let's open an issue, as that is a battle for another day...

**Comment by pllim** [comment]
I removed the numpy-dev label because it is technically not "dev" anymore.

File: docs/io/fits/usage/unfamiliar.rst (hunk 2 above)
Comment (drdavella) [reply docs/io/fits/usage/unfamiliar.rst]: It's not a bug, it's just a missing feature: https://github.com/astropy/pytest-doctestplus/issues/18

**Comment by mhvk** [comment]
Makes sense. Please do open issues!

**Comment by mhvk** [comment]
@drdavella - the sphinx build is failing, not quite sure why....

**Comment by mhvk** [comment]
Oops, sorry, I guess that's why you have #7257...

**Comment by bsipocz** [comment]
All green, so I'm merging. Thank you @mhvk and @drdavella!

Everyone else, please rebase your PRs.

**Pull request merged**